third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -463,574 +391,1047 @@ Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+    Value fatPtrBase = fatPtr[0];
+    Value fatPtrOffset = fatPtr[1];
+    Location curLoc = storeOp.getLoc();
+
+    llvm::SmallDenseMap<StringAttr, Attribute> attributes{};
+    Value newPtr = fatPtrBase;
+    if (llvm::isa<RankedTensorType>(storeOp.getPtr().getType())) {
+      const FatPointers::FatPtrAttrs &fatPtrAttrs =
+          fatPtrs.at({fatPtrBase, fatPtrOffset});
+      newPtr = createTensorPointer(rewriter, fatPtrBase, fatPtrOffset, curLoc,
+                                   fatPtrAttrs);
+    }
+    SmallVector<Value> operands =
+        storeOp.getOperands().take_back(storeOp.getNumOperands() - 1);
+    operands.insert(operands.begin(), newPtr);
+    SmallVector<NamedAttribute> attrs = llvm::to_vector(storeOp->getAttrs());
+    rewriter.replaceOpWithNewOp<tt::StoreOp>(storeOp, TypeRange{},
+                                             ValueRange{operands}, attrs);
+    return success();
+  }
+};
+
+static const std::string kInitFuncArgsRewritten =
+    kPtrCanonPrefix + "init-func-ptr-args";
+/// tt.func gets rewritten differently from all the other ops - the op itself is
+/// not rewritten. What is rewritten are all tt.ptr args are rewritten (all
+/// uses) to be %1 = unrealize_cast(%arg0: tt.ptr, c0: i32) -> tt.ptr. This
+/// unrealized_cast is then (possibly) materialized in the second pass
+/// (ConvertUnimplementedOpUnrealizedCasts) if it wasn't DCEd (via a user
+/// extracting the tt.ptr and c0 operands).
+struct InitFuncPtrArgs : OpRewritePattern<tt::FuncOp> {
+  InitFuncPtrArgs(MLIRContext *context, FatPointers &fatPtrs)
+      : OpRewritePattern(context, 0), fatPtrs(fatPtrs) {}
+
+  LogicalResult matchAndRewrite(tt::FuncOp newOp,
+                                PatternRewriter &rewriter) const override {
+    if (newOp->hasAttr(kInitFuncArgsRewritten))
       return failure();
 
-    // Clean-up: don't assume the operation to delete are in the correct order,
-    // but force dropping the reference of the ops before we delete them
-    for (Operation *op : opToDelete) {
-      op->dropAllReferences();
-      op->dropAllDefinedValueUses();
-      rewriter.eraseOp(op);
+    int64_t bitness = 64;
+    rewriter.setInsertionPointToStart(&newOp.getBody().front());
+    for (auto [idx, arg] : llvm::enumerate(newOp.getArguments())) {
+      // The pointer argument needs to be a scalar
+      if (!isa<tt::PointerType>(arg.getType()))
+        continue;
+      if (auto pointerRangeAttr =
+              newOp.getArgAttrOfType<IntegerAttr>(idx, "tt.pointer_range"))
+        bitness = pointerRangeAttr.getInt();
+      Value zeroOffset =
+          rewriter.create<arith::ConstantIntOp>(newOp.getLoc(), 0, bitness);
+      auto dummyCast = rewriter.create<UnrealizedConversionCastOp>(
+          arg.getLoc(), TypeRange{arg.getType()}, ValueRange{arg, zeroOffset});
+      rewriter.replaceAllUsesExcept(arg, dummyCast.getResult(0), dummyCast);
+      // TODO(max): why is this true?
```

Comment:
This is effectively the initial state for `canNarrow`? @giuseros

Reply:
Yea @giuseros already responded to most of these TODO questions above